### ISSUE TITLE
data sync options for STORE

### ISSUE THREAD

**Issue by tomcucinotta** [tomcucinotta/distwalk#21]
The STORE command has been sketched out with the idea of a per-request sync, with the ODIRECT option specifiable for the whole node as command-line option to dw_node.
We'd need to add more options in this regard, to mimic common syncing behavior of a multitude of systems (ideally, still under control of the client):
- no syncing at all
- per-request data sync
- periodic syncing of written changes (e.g., common of DBs, MongoDB/WiredTiger, Cassandra, ...) with customizable period

**Comment by tomcucinotta** [comment]
still missing the "no sync" option, to close this; not sure if it makes sense to distinguish:
- writes that need an immediate sync before getting a response
- writes that do not sync immediately, but will wait for the next periodic sync to have been completed, before getting a response (i.e., when the periodic sync happens, we need to track what req_ids are being synced)
- writes that simply write() and respond immediately to the client.

**Comment by deRemo** [comment]
Current behavior:

if periodic sync is NOT specified:
  STORE N,nosync -> store N bytes and no fsync()
  STORE N,sync -> store N bytes, then fsync()

if periodic_sync_msec is specified:
  STORE N,nosync -> same as before
  STORE N,sync -> store N bytes, then wait for the periodic sync timer to trigger fsync()

**Comment by tomcucinotta** [comment]
in the current code, below, if the periodic sync is used, and wait_sync option specified, then the read() blocks the storage thread until the periodic sync; this is not what I had in mind. Instead "wait for the next periodic sync to have been completed, before getting a response", means that I'll keep doing additional write()s as they arrive, but only when the periodic sync comes, I'll let all the requests with a store since the last sync, to move forward in the sequence of commands within the messages they belong to, and send responses to the client.
Moreover, the response should be sent once the sync is done, but at the moment this is NOT the case: the storage worker needs to go back to its epoll() loop, to experience the (type == TIMER) case, and trigger the fsync(). However, here the store() function waits till the timer, then completes, causing the caller (storage_worker) to relpy to the worker thread in dw_node, that replies to the client while the fsync() is still in progress.
```
    if (wait_sync) { // wait for periodic timer to be triggered
        if (infos->periodic_sync_msec > 0) {
            uint64_t val;
            if (read(infos->timerfd, &val, sizeof(uint64_t)) < 0) {
                perror("periodic sync read()");
            }
        } else {
            fsync(storage_info->storage_fd);
        }
    }
```

**Comment by tomcucinotta** [comment]
In other words, the wait_sync option with period sync, as I have it in mind, requires introducing a queue of pending STORE requests within the storage thread, where requests are appended, till the next periodic sync comes. At that moment, after the fsync() completes, we should check the pending STORE requests, and reply to their worker threads.

**Comment by deRemo** [comment]
fixed in e21d707
